277 pipeline editor easier search in the step chooser left sidebar

sivabalansm opened the pull request:
Added search bar that searches through files and metadata of scripts and files, highlights what it found and ranks them based on relevant to the query. Also added foldable directories for scripts and pipelines. 
closes #277

sivabalansm committed:
Add search bar to stepchooser, filter out pipelines and scripts based on query

sivabalansm committed:
Add highlighting of searched terms in search feature

sivabalansm committed:
Add foldable directories for pipeliens and scripts

jmlord commented on ui/src/components/PipelineEditor/StepChooser.jsx:
```diff
@@ -18,6 +19,161 @@ const onDragStart = (event, nodeType, descriptionFile) => {
   event.dataTransfer.effectAllowed = "move";
 };
 
+// I hope jean micheal enjoys this code review :D, comments galore!
```
Remove this comment, hahaha!
:)

jmlord commented on ui/src/components/PipelineEditor/StepChooser.jsx:
```diff
@@ -55,10 +211,55 @@ function PipelineStep({ descriptionFile, fileName, selectedStep, stepName, onSte
   );
 }
 
+function SearchResultStep({ descriptionFile, fileName, selectedStep, stepName, onStepClick, searchQuery, metadata }) {
```
The file is getting long, we could split these new classes.

jmlord commented on ui/src/components/PipelineEditor/StepChooser.jsx:
```diff
@@ -18,6 +19,161 @@ const onDragStart = (event, nodeType, descriptionFile) => {
 
+// I hope jean micheal enjoys this code review :D, comments galore!
+// Helper function to highlight text with search keywords
+function highlightText(text, searchQuery) {
```
On dirait presque un exercice d'école. A+

jmlord commented on ui/src/components/PipelineEditor/StepChooser.jsx:
```diff
@@ -18,6 +19,161 @@ const onDragStart = (event, nodeType, descriptionFile) => {
+    parts.push(text.substring(lastIndex));
+  }
+
+  return parts.length > 0 ? parts : text;
```
parts.length > 0 is always true because of 

```
if (lastIndex < text.length) {
    parts.push(text.substring(lastIndex));
  }
```
If there are no matches this will put the whole text.

jmlord committed:
Cleanup unused vars and trailing whitespaces.

jmlord committed:
Small styling adjustments

jmlord committed:
moved Siva's highlightText to utils

jmlord committed:
fix highlight was adding a space in the words

jmlord reviewed (approved):
Thanks for this last contribution!
I'll do any necessary adjustments.

PS.: You should add an extension that removes trailing whitespaces when you save.

jmlord committed:
this part was redundant with the next loop.

jmlord committed:
Checking metadata parts individually
+ Change ellipsis logic
+ Split to keywords before splitting the execution flow in two.

jmlord committed:
no need to fetch if we already have the metadata.
Normally this is handled by the fetch call, but in this case we wanted a delay to avoid blocking more important requests.

jmlord committed:
searchKeywords as a state

jmlord committed:
Key already on SearchResultStep
no need to pass file name
+ keep highlight in state

jmlord committed:
Pre-loading done by PipelineStep
No separate preloading structure needed

jmlord committed:
Metadata search funcitons in separate file.
+ using search keywords instead of full query
+ removed matchType, relying only on score

jmlord committed:
lowercase was not always done
doing it upfront

jmlord committed:
Score on number of matched keywords

jmlord committed:
count keywords in metadata too
Using approx with stringify

jmlord committed:
moved inline functions to proper functions

jmlord committed:
single array of results
+ visual difference between scripts and pipelines

jmlord committed:
Added magnifying glass icon

jmlord committed:
Documented object structure
+ score not usefull after sort.

jmlord committed:
SearchResultStep in StepSearch file
along with search functions

jmlord committed:
moved function to outer scope

jmlord committed:
Autofocus on search field
